samples/web/content/apprtc/gcm_register.py

```diff
@@ -0,0 +1,222 @@
+"""
+
+import constants
+import datetime
+import json
+import logging
+import webapp2
+
+from google.appengine.ext import ndb
+
+def get_ancestor_key():
+  return ndb.Key('GCMRecord', 'global')
+
+PARAM_USER_ID = 'userId'
+PARAM_GCM_ID = 'gcmId'
+PARAM_CODE = 'code'
+PARAM_OLD_GCM_ID = 'oldGcmId'
+PARAM_NEW_GCM_ID = 'newGcmId'
+PARAM_USER_ID_LIST = 'userIdList'
+
+class BindPage(webapp2.RequestHandler):
+  def get_message_from_json(self):
+    try:
+      message = json.loads(self.request.body)
+      return message
+    except Exception as e:
+      logging.warning('JSON load error from BindPage: error=' + str(e)
+          + ', request=' + self.request.body)
+      return None
+
+  def handle_new(self):
+    msg = self.get_message_from_json()
+    if msg and msg[PARAM_USER_ID] and msg[PARAM_GCM_ID]:
+      # TODO(jiayl): validate the input, generate a random code, and send SMS.
+      self.response.out.write(GCMRecord.add_or_update(
+          msg[PARAM_USER_ID], msg[PARAM_GCM_ID], 'fake_code'))
+    else:
+      self.response.out.write(constants.RESPONSE_INVALID_ARGUMENT)
+
+  def handle_update(self):
+    msg = self.get_message_from_json()
+    if msg and msg[PARAM_USER_ID] \
+        and msg[PARAM_OLD_GCM_ID] and msg[PARAM_NEW_GCM_ID]:
+      self.response.out.write(GCMRecord.update_gcm_id(
+          msg[PARAM_USER_ID], msg[PARAM_OLD_GCM_ID], msg[PARAM_NEW_GCM_ID]))
+    else:
+      self.response.out.write(constants.RESPONSE_INVALID_ARGUMENT)
+
+  def handle_verify(self):
+    msg = self.get_message_from_json()
+    if msg and msg[PARAM_USER_ID] and msg[PARAM_GCM_ID] and msg[PARAM_CODE]:
+      self.response.out.write(GCMRecord.verify(
+          msg[PARAM_USER_ID], msg[PARAM_GCM_ID], msg[PARAM_CODE]))
+    else:
+      self.response.out.write(constants.RESPONSE_INVALID_ARGUMENT)
+
+  def handle_query(self):
+    msg = self.get_message_from_json()
+    if msg.has_key(PARAM_USER_ID_LIST) and msg[PARAM_USER_ID_LIST]:
+      result = []
+      for id in msg[PARAM_USER_ID_LIST]:
+        # TODO(jiayl): Only return the verified users when SMS verification is
+        # added.
+        if len(GCMRecord.get_by_user_id(id)) > 0:
+          result.append(id)
+      self.response.out.write(json.dumps(result))
+    elif msg.has_key(PARAM_GCM_ID) and msg[PARAM_GCM_ID]:
+      records = GCMRecord.get_by_gcm_id(msg[PARAM_GCM_ID])
+      assert len(records) < 2
+      if len(records) == 0:
+        self.response.out.write(constants.RESPONSE_NOT_FOUND)
+      elif records[0].verified:
+        self.response.out.write(constants.RESPONSE_VERIFIED)
+      else:
+        self.response.out.write(constants.RESPONSE_UNVERIFIED)
+    else:
+      self.response.out.write(constants.RESPONSE_INVALID_ARGUMENT)
+
+  def handle_del(self):
+    msg = self.get_message_from_json()
+    if msg and msg[PARAM_USER_ID] and msg[PARAM_GCM_ID]:
+      GCMRecord.remove(msg[PARAM_USER_ID], msg[PARAM_GCM_ID])
+      self.response.out.write(constants.RESPONSE_SUCCESS)
```

Comment:
Do we want to return SUCCESS in the case where we didn't find a matching record?